// @heroku-cli/command/lib/command.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Command = void 0;
const tslib_1 = require("tslib");
const core_1 = require("@oclif/core");
const util_1 = require("util");
const yargs_parser_1 = tslib_1.__importDefault(require("yargs-parser"));
const yargs_unparser_1 = tslib_1.__importDefault(require("yargs-unparser"));
const pjson = require('../package.json');
const deps_1 = tslib_1.__importDefault(require("./deps"));
const deprecatedCLI = (0, util_1.deprecate)(() => {
    return require('cli-ux').cli;
}, 'this.out and this.cli is deprecated. Please import "CliUx" from the @oclif/core module directly instead.');
class Command extends core_1.Command {
    constructor() {
        super(...arguments);
        this.base = `${pjson.name}@${pjson.version}`;
        this.allowArbitraryFlags = false;
    }
    get heroku() {
        if (this._heroku)
            return this._heroku;
        this._heroku = new deps_1.default.APIClient(this.config);
        return this._heroku;
    }
    get legacyHerokuClient() {
        if (this._legacyHerokuClient)
            return this._legacyHerokuClient;
        const HerokuClient = require('heroku-client');
        const options = {
            debug: this.config.debug,
            host: `${this.heroku.defaults.protocol || 'https:'}//${this.heroku.defaults.host ||
                'api.heroku.com'}`,
            token: this.heroku.auth,
            userAgent: this.heroku.defaults.headers['user-agent'],
        };
        this._legacyHerokuClient = new HerokuClient(options);
        return this._legacyHerokuClient;
    }
    get cli() {
        return deprecatedCLI();
    }
    get out() {
        return deprecatedCLI();
    }
    async parse(options, argv) {
        if (this.allowArbitraryFlags) {
            try {
                return await super.parse(options, argv);
            }
            catch (error) {
                const { flags: nonExistentFlags } = error;
                const parsed = (0, yargs_parser_1.default)(this.argv);
                const nonExistentFlagsWithValues = Object.assign({}, parsed);
                if (nonExistentFlags && nonExistentFlags.length > 0) {
                    this.warn(`Using [${nonExistentFlags}] without a '--' (end of options) preceeding them is deprecated. Please use '--' preceeding the flag(s) meant to be passed-though.`);
                    for (const flag of nonExistentFlags) {
                        const key = flag.replace('--', '');
                        delete parsed[key];
                    }
                }
                for (const key in parsed) {
                    if (Reflect.has(parsed, key)) {
                        delete nonExistentFlagsWithValues[key];
                    }
                }
                this.argv = (0, yargs_unparser_1.default)(parsed);
                const result = await super.parse(options, argv);
                result.nonExistentFlags = (0, yargs_unparser_1.default)(nonExistentFlagsWithValues);
                return result;
            }
        }
        return super.parse(options, argv);
    }
}
exports.Command = Command;
